src/Controller/LoginController.php

```diff
@@ -0,0 +1,438 @@
+
+    /** @var Simple  */
+    protected Simple $authSource;
+
+    /** @var Utils\HTTP */
+    protected Utils\HTTP $httpUtils;
+
+    /** @var Cas20 */
+    protected Cas20 $cas20Protocol;
+
+    /** @var TicketStore */
+    protected TicketStore $ticketStore;
+
+    /** @var ServiceValidator */
+    protected ServiceValidator $serviceValidator;
+
+    /** @var array */
+    protected array $idpList;
+
+    /** @var string|null */
+    protected ?string $authProcId = null;
+
+    protected array $postAuthUrlParameters = [];
+
+    /** @var string[] */
+    private const DEBUG_MODES = ['true', 'samlValidate'];
+
+    /** @var AttributeExtractor */
+    protected AttributeExtractor $attributeExtractor;
+
+    /** @var SamlValidateResponder */
+    private SamlValidateResponder $samlValidateResponder;
+
+    /**
+     * @param   Configuration       $sspConfig
+     * @param   Configuration|null  $casConfig
+     * @param   Simple|null         $source
+     * @param   Utils\HTTP|null     $httpUtils
+     *
+     * @throws \Exception
+     */
+    public function __construct(
+        private readonly Configuration $sspConfig,
+        // Facilitate testing
+        Configuration $casConfig = null,
+        Simple $source = null,
+        Utils\HTTP $httpUtils = null,
+    ) {
+        $this->casConfig = ($casConfig === null || $casConfig === $sspConfig)
+            ? Configuration::getConfig('module_casserver.php') : $casConfig;
+        // Saml Validate Responsder
+        $this->samlValidateResponder = new SamlValidateResponder();
+        // Service Validator needs the generic casserver configuration.
+        $this->serviceValidator = new ServiceValidator($this->casConfig);
+        $this->authSource = $source ?? new Simple($this->casConfig->getValue('authsource'));
+        $this->httpUtils = $httpUtils ?? new Utils\HTTP();
+    }
+
+    /**
+     *
+     * @param   Request      $request
+     * @param   bool         $renew
+     * @param   bool         $gateway
+     * @param   string|null  $service
+     * @param   string|null  $TARGET  Query parameter name for "service" used by older CAS clients'
+     * @param   string|null  $scope
+     * @param   string|null  $language
+     * @param   string|null  $entityId
+     * @param   string|null  $debugMode
+     * @param   string|null  $method
+     *
+     * @return RunnableResponse
+     */
+    public function login(
+        Request $request,
+        #[MapQueryParameter] bool $renew = false,
+        #[MapQueryParameter] bool $gateway = false,
+        #[MapQueryParameter] ?string $service = null,
+        #[MapQueryParameter] ?string $TARGET = null,
+        #[MapQueryParameter] ?string $scope = null,
+        #[MapQueryParameter] ?string $language = null,
+        #[MapQueryParameter] ?string $entityId = null,
+        #[MapQueryParameter] ?string $debugMode = null,
+        #[MapQueryParameter] ?string $method = null,
+    ): RunnableResponse {
```

Comment:
I think return type of Runnable Response is to strict since Debug mode should return html content.

Reply:
done